File: client1.js
const net = require('net');
const fs = require('fs');
const WebSocket = require('ws');


const socketPath = '/tmp/node_cpp_socket.sock';
let JsonData = {
            "call_stat": 0,
            "call_req": 0,
            "call_acc": 0,
            "seat_alm": 0,
            "signal": 0
        };
let responseJson = JSON.stringify(JsonData)

if (fs.existsSync(socketPath)) {
  fs.unlinkSync(socketPath);
}

const server = net.createServer((socket) => {
  console.log('C++ クライアントが接続しました。');

  socket.on('data', (data) => {
    try {
      const receivedData = JSON.parse(data.toString());
      console.log('C++側から受信したデータ:', receivedData);

    //   const responseData = { message: 'Hello from Node.js' };
      const responseData = receivedData;
      responseJson = JSON.stringify(responseData);

      socket.write(responseJson);
    } catch (error) {
      console.error('データのパースに失敗しました:', error.message);
    }
  });

  socket.on('close', () => {
    console.log('クライアントが切断されました。');
  });

  socket.on('error', (err) => {
    console.error('エラーが発生しました:', err.message);
  });
});

server.listen(socketPath, () => {
  console.log(`Node.js サーバーが ${socketPath} で待機中...`);
});

//websocketServer側
const wss = new WebSocket.Server({ port:5002 });
wss.on('connection', (ws) => {
    // WebSocket接続が確立されたときの処理
    console.log('WebSocket connected');
  
    // サンプルJSONデータ
    // const jsonData = { message: 'Hello from the Nodejs server!' };
    // const jsonData = responseData;
  
    // JSONデータを文字列に変換してWebSocketへ送信
    // ws.send(JSON.stringify(jsonData));
    // ws.send(JSON.stringify(responseData));
    ws.send(responseJson)//サーバーから一発目送信
  // WebSocketからメッセージを受信したときの処理
  ws.on('message', (message) => {
    // 受信したJSONデータをパース
    const receivedData = JSON.parse(message);
    
    // データを加工または変更
    receivedData.message = 'Server modified: ' + receivedData.message;

    // 加工したデータを文字列に変換してクライアントに送り返す
    // ws.send(JSON.stringify(receivedData));
    setTimeout(() => {
      ws.send(responseJson);
    }, 100);
  });
});

File: client1.cpp
#include <iostream>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#include <cstring>
#include <cstdio>
#include <nlohmann/json.hpp>
#include <csignal>
#include <thread>
#include <chrono>

using json = nlohmann::json;

bool keepRunning = true;

// 受信したJSONデータを保持する変数
json receivedData;
// 送信するJSONデータの作成
json sendData = {
    // {"message", "Hello from C++"},
    // {"key", "new_value"},  // データを変更
    // 他のデータフィールドを追加できます
    {"call_stat", 0},
    {"call_req", 0},
    {"call_acc", 0},
    {"seat_alm", 0},
    {"signal", 0}
};

// Ctrl+C でプログラムを終了するためのシグナルハンドラ
void handleSignal(int signal) {
    if (signal == SIGINT) {
        std::cout << "Ctrl+Cが押されました。プログラムを終了します。\n";
        keepRunning = false;
    }
}

// 新しいスレッドで実行される関数
void threadFunction() {
    const char *socketPath = "/tmp/node_cpp_socket.sock";

    int sockfd = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sockfd == -1) {
        perror("ソケットの作成に失敗しました");
        return;
    }

    sockaddr_un serverAddr{};
    serverAddr.sun_family = AF_UNIX;
    strncpy(serverAddr.sun_path, socketPath, sizeof(serverAddr.sun_path) - 1);

    if (connect(sockfd, reinterpret_cast<sockaddr *>(&serverAddr), sizeof(serverAddr)) == -1) {
        perror("サーバーへの接続に失敗しました");
        close(sockfd);
        return;
    }

    while (keepRunning) {
        static int cnt=0;
        // JSONデータを文字列に変換
        std::string jsonString = sendData.dump();

        // データをサーバーに送信
        if (write(sockfd, jsonString.c_str(), jsonString.length()) == -1) {
            perror("データの送信に失敗しました");
            close(sockfd);
            return;
        }

        char buffer[1024];
        ssize_t bytesRead = read(sockfd, buffer, sizeof(buffer) - 1);
        if (bytesRead == -1) {
            perror("データの受信に失敗しました");
            close(sockfd);
            return;
        }

        buffer[bytesRead] = '\0';
        std::cout << "サーバーからのレスポンス: " << buffer << std::endl;

        // 受信したJSONデータを保持
        receivedData = json::parse(buffer);

        // 受信したデータを次に送信するデータに使用（例：keyフィールドを変更）
        sendData = receivedData; 
        sendData["call_stat"] = cnt;
        sendData["seat_alm"] = cnt;
        // sendData[""] = cnt;

        // 一定の待機時間を挿入（必要に応じて調整）
        //sleep(2);
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        cnt++;
    }

    close(sockfd);
}

int main() {
    // Ctrl+C でプログラムを終了するためのシグナルハンドラの設定
    signal(SIGINT, handleSignal);

    // 新しいスレッドを開始
    std::thread workerThread(threadFunction);

    // メインスレッドはここで待機（Ctrl+Cが押されるまで）
    workerThread.join();

    return 0;
}